pkg/helm/charts.go

```diff
@@ -0,0 +1,289 @@
+func NewRepoManager(kc client.Client, cacheDir string) *RepoManager {
+	return &RepoManager{
+		Client:   kc,
+		CacheDir: cacheDir,
+	}
+}
+
+// RepoManager implements HelmRepoManager interface using the Helm library packages.
+type RepoManager struct {
+	client.Client
+	CacheDir string
+}
+
+// ChartReference is a Helm chart reference, the SourceRef is a Flux
+// SourceReference for the Helm chart.
+type ChartReference struct {
+	Chart     string
+	Version   string
+	SourceRef helmv2beta1.CrossNamespaceObjectReference
+}
+
+// DefaultChartGetter provides default ways to get a chart index.yaml based on
+// the URL scheme.
+var DefaultChartGetters = getter.Providers{
+	getter.Provider{
+		Schemes: []string{"http", "https"},
+		New:     getter.NewHTTPGetter,
+	},
+}
+
+type ChartPredicate func(*repo.ChartVersion) bool
+
+// Profiles is a predicate for scanning charts with the ProfileAnnotation.
+var Profiles = func(v *repo.ChartVersion) bool {
+	return hasAnnotation(v.Metadata, ProfileAnnotation)
+}
+
+// ScanCharts filters charts using the provided predicate.
+// TODO: Add caching based on the Status Artifact Revision.
+func (h *RepoManager) GetCharts(ctx context.Context, hr *sourcev1beta1.HelmRepository, pred ChartPredicate) ([]*pb.Profile, error) {
+	chartRepo, err := fetchIndexFile(hr.Status.URL)
+	if err != nil {
+		return nil, fmt.Errorf("fetching profiles from HelmRepository %s/%s %q: %w",
+			hr.GetName(), hr.GetNamespace(), hr.Spec.URL, err)
+	}
+
+	ps := make(map[string]*pb.Profile)
+
+	for name, versions := range chartRepo.Entries {
+		for _, v := range versions {
+			if pred(v) {
+				// if already added, update the versions array
+				if p, ok := ps[name]; ok {
+					p.AvailableVersions = append(p.AvailableVersions, v.Version)
+				} else { // otherwise create a new profile and add to map
+					p = &pb.Profile{
+						Name:        name,
+						Home:        v.Home,
+						Sources:     v.Sources,
+						Description: v.Description,
+						Keywords:    v.Keywords,
+						Icon:        v.Icon,
+						KubeVersion: v.KubeVersion,
+					}
+					for _, m := range v.Maintainers {
+						p.Maintainers = append(p.Maintainers, &pb.Maintainer{
+							Name:  m.Name,
+							Email: m.Email,
+							Url:   m.URL,
+						})
+					}
+					p.AvailableVersions = append(p.AvailableVersions, v.Version)
+					ps[name] = p
+				}
+			}
+		}
+	}
+
+	profiles := []*pb.Profile{}
+
+	for _, p := range ps {
+		sort.Strings(p.AvailableVersions)
+		profiles = append(profiles, p)
+	}
+
+	return profiles, nil
+}
+
+// GetValuesFile fetches the value file from a chart.
+func (h *RepoManager) GetValuesFile(ctx context.Context, helmRepo *sourcev1beta1.HelmRepository, c *ChartReference, filename string) ([]byte, error) {
+	if err := h.updateCache(ctx, helmRepo); err != nil {
+		return nil, err
+	}
+
+	chart, err := h.loadChart(ctx, helmRepo, c)
+	if err != nil {
+		return nil, fmt.Errorf("loading %s from chart: %w", filename, err)
+	}
```

Comment:
Given you do not seem to validate the chart (`Validate` method on the object), and the minimal validation Helm does while loading can be easily replicated and is not likely to change, I think it would be much more memory efficient to not load the whole data into memory but rather scan through the tarball for the `chartutil.ValuesfileName` in the root of the archive.

We do something similar in a part of the source-controller for the metadata file: https://github.com/fluxcd/source-controller/blob/59dc602b2d0a5b0d4625a4a848303fa2d2117e84/internal/helm/chart/metadata.go#L142

Reply:
:thinking: that change sounds reasonable, but I quite like that we are delegating a lot of the complexity of interacting with the chart to the Helm library. I think for the moment this change might be a bit overkill, but in this [story](https://github.com/weaveworks/weave-gitops/issues/1144) we are getting into caching (and in-general being more optimised), I'm going to flag your comment in that issue :smile:

Reply:
The more detailed explanation for this advise is Helm umbrella charts. When Helm loads such a chart, it also includes all subcharts because their inclusion for the apply is determined based on the values used during release. This means that you load a whole lot tree of nonsense while you are after very specific data.